### PR TITLE
chore: Upgrades app to modern Typescript

### PR DESCRIPTION
**Pull request by markrickert**
# Summary

This PR converts the entire app to modern typescript. Along with this comes a lot of package upgrades now that `react-diff-view` and `framer-motion` support typescript

This addresses the Typescript support item in #108

## Test Plan

* `yarn test` runs and passes
* I added a new `yarn typecheck` command and added it to the github workflow
* I couldn't get the docker image up and running on my local M2 machine, so `yarn test-e2e` is failing locally for me.

## Checklist

- [x] I tested this thoroughly
- [x] I added the documentation in `README.md` (if needed)

### CODE REVIEW

**Comment by pvinis** [comment]
amazing! we've been wanting to move it to TS for a while now.

Allow me a couple of days because I'm on holidays, and I will review it on Wednesday 👍.

**Comment by markrickert** [comment]
No worries. The conversion to typescript found quite a few small bugs... typos, prop type mismatches, etc... so i fixed them! I've `@ts-ignore`d the places I couldn't get the linter to pass so it might be worth visiting those places in the code.

**Review by pvinis — APPROVED** [review]
looks great!

**Comment by pvinis** [comment]
one tiny thing, i saw that the icons are not showing until hover.
Screenshot 2024-02-07 at 19 57 58
Screenshot 2024-02-07 at 19 58 04
fix that, and then we are good to go!

**Review by pvinis — APPROVED** [review]
changing to avoid accidental merge

**Review by pvinis — CHANGES_REQUESTED** [review]
changing to avoid accidental merge

File: src/releases/react-native/0.72.tsx [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import React, { Fragment } from 'react'
-import { ReleaseT } from '../types'
+import type { ReleaseT } from '../types'
```
Comment (pvinis) [inline src/releases/react-native/0.72.tsx:2]: is that needed? ts is smart enough to know this, right?

Comment (markrickert) [reply src/releases/react-native/0.72.tsx:2]: Just a style preference. The TS compiler knows the difference. 

While week on the topic, want are your thoughts on using the linter to enforce import order?

**Pull request merged**